usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -0,0 +1,184 @@
+        "filter_queryset_func": subaward_filter,
+    },
+}
+
+
+@api_transformations(api_version=settings.API_VERSION, function_list=API_TRANSFORM_FUNCTIONS)
+class SpendingByAwardVisualizationViewSet(APIView):
+    """
+    This route takes award filters and fields, and returns the fields of the filtered awards.
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/search/AdvancedSearch.md"
+
+    @cache_response()
+    def post(self, request):
+        """Return all awards matching the provided filters and limits"""
+        json_request = self.validate_request_data(request.data)
+        self.is_subaward = json_request["subawards"]
+        self.constants = GLOBAL_MAP["subaward"] if self.is_subaward else GLOBAL_MAP["award"]
+        self.filters = json_request["filters"]
+        self.fields = json_request["fields"]
+        self.pagination = {
+            "limit": json_request["limit"],
+            "lower_bound": (json_request["page"] - 1) * json_request["limit"],
+            "page": json_request["page"],
+            "sort_key": json_request.get("sort") or self.fields[0],
+            "sort_order": json_request["order"],
+            "upper_bound": json_request["page"] * json_request["limit"] + 1,
+        }
+
+        if self.if_no_intersection():  # Like an exception, but API response is a HTTP 200 with a JSON payload
+            return Response(self.populate_response(results=[], has_next=False))
+        raise_if_award_types_not_valid_subset(self.filters["award_type_codes"], self.is_subaward)
+        raise_if_sort_key_not_valid(self.pagination["sort_key"], self.fields, self.is_subaward)
+
+        return Response(self.create_response(self.construct_queryset()))
+
+    @staticmethod
+    def validate_request_data(request_data):
+        models = [
+            {"name": "fields", "key": "fields", "type": "array", "array_type": "text", "text_type": "search", "min": 1},
+            {"name": "subawards", "key": "subawards", "type": "boolean", "default": False},
+        ]
+        models.extend(copy.deepcopy(AWARD_FILTER))
+        models.extend(copy.deepcopy(PAGINATION))
+        for m in models:
+            if m["name"] in ("award_type_codes", "fields"):
+                m["optional"] = False
+
+        return TinyShield(models).block(request_data)
+
+    def if_no_intersection(self):
+        # "Special case" behavior: there will never be results when the website provides this value
+        return "no intersection" in self.filters["award_type_codes"]
+
+    def construct_queryset(self):
+        sort_by_fields = self.get_sort_by_fields()
+        database_fields = self.get_database_fields()
+        base_queryset = self.constants["filter_queryset_func"](self.filters)
+        queryset = self.custom_queryset_order_by(base_queryset, sort_by_fields, self.pagination["sort_order"])
+        return queryset.values(*list(database_fields))[self.pagination["lower_bound"] : self.pagination["upper_bound"]]
```

Comment:
Does Django magic ensure that we only fetch the ones we want, or are we doing the whole query here and just slicing off the chunk we want to show?

Reply:
Queryset are lazy loaded, the DB still hasn't been queried yet. This is actually modifying the queryset to add `limit` and `offset` if necessary